build/item.rs

```diff
@@ -0,0 +1,170 @@
+        .map(|i| {
+            let item = ident(i.name.to_pascal_case());
+            let translation_key = &i.translation_key;
+            quote! {
+                Self::#item => #translation_key,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_varients = items
+        .iter()
+        .map(|i| ident(i.name.to_pascal_case()))
+        .collect::<Vec<_>>();
+
+    let item_from_str_arms = items
+        .iter()
+        .map(|i| {
+            let name = &i.name;
+            let name_ident = ident(name.to_pascal_case());
+            quote! {
+                #name => Some(Item::#name_ident),
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_to_str_arms = items
+        .iter()
+        .map(|i| {
+            let name = &i.name;
+            let name_ident = ident(name.to_pascal_case());
+            quote! {
+                Item::#name_ident => #name,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_to_raw_id_arms = items
+        .iter()
+        .map(|i| {
+            let id = &i.id;
+            let name_ident = ident(&i.name.to_pascal_case());
+
+            quote! {
+                Item::#name_ident => #id,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_from_raw_id_arms = items
+        .iter()
+        .map(|i| {
+            let id = &i.id;
+            let name_ident = ident(&i.name.to_pascal_case());
+
+            quote! {
+                #id => Some(Item::#name_ident),
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_to_max_count_arms = items
+        .iter()
+        .map(|i| {
+            let name_ident = ident(&i.name.to_pascal_case());
+            let max_count = i.max_stack;
+
+            quote! {
+                Item::#name_ident => #max_count,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let items_varients = items
+        .iter()
+        .map(|i| ident(i.name.to_pascal_case()))
+        .collect::<Vec<_>>();
+
+    Ok(quote! {
+        /// Represents an item from game
+        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+        pub enum Item {
+            #(#item_varients,)*
+        }
+
+        impl Item {
+            /// Constructs a item from a raw item ID.
+            ///
+            /// If the given ID is invalid, `None` is returned.
+            pub const fn from_raw(id: u16) -> Option<Self> {
+                match id {
+                    #item_from_raw_id_arms
+                    _ => None
+                }
+
+                //if id <= #items_count.try_into().unwrap() {
+                //    Some(Self(id))
+                //} else {
+                //    None
+                //}
+            }
+
+            /// Gets the raw item ID from the item
+            pub const fn to_raw(self) -> u16 {
+                match self {
+                    #item_to_raw_id_arms
+                }
```

Comment:
If we make the discriminant of the `Item` enum be the item ID, we can implement this function as
```rust
self as u16
```

Then the definition of `Item` would look something like this. (If you know the IDs are sequential starting at 0 you can skip this part. But I think it would be good style to do this anyway)

```rust
enum {
    // (idk what the actual item IDs are)
    Apple = 0,
    Stone = 1,
    Potato = 2,
    ...
}
```

The implementation of `from_raw` would stay the same.

See also: https://doc.rust-lang.org/reference/items/enumerations.html#custom-discriminant-values-for-fieldless-enumerations

Reply:
Yeah, the item ids are sequential form 0, so should I still add the item ids in the enum? I have implemented it now where there's no item ids in the enum, but `self as u16` works.

Reply:
Yea you can leave the item IDs off.